### PR TITLE
CPDP-1215: DMS File Submission

### PR DESCRIPTION
**Pull request by mikail-khan**

### CODE REVIEW

File: app/uk/gov/hmrc/cgtpropertydisposals/service/DmsSubmissionService.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import configs.syntax._
+import play.api.Configuration
+import uk.gov.hmrc.cgtpropertydisposals.connectors.GFormConnector
+import uk.gov.hmrc.cgtpropertydisposals.models.Error
+import uk.gov.hmrc.cgtpropertydisposals.models.dms._
+import uk.gov.hmrc.cgtpropertydisposals.models.ids.CgtReference
+import uk.gov.hmrc.cgtpropertydisposals.util.Logging
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/DmsSubmissionService.scala]: you can inject an execution context in

File: app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    cgtReference: CgtReference,
+    fileDescriptorId: FileDescriptorId
+  ): EitherT[Future, Error, Option[UpscanFileDescriptor]] =
+    upscanFileDescriptorRepository.get(fileDescriptorId)
+
+  override def downloadFilesFromS3(
+    snapshot: UpscanSnapshot,
+    urls: List[UpscanCallBack]
+  ): EitherT[Future, Error, List[Either[Error, FileAttachment]]] = {
+    implicit val cs: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala]: could you not use the injected execution context?

File: app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  def getUpscanFileDescriptor(fileDescriptorId: FileDescriptorId): EitherT[Future, Error, Option[UpscanFileDescriptor]]
+
+  def updateUpscanFileDescriptorStatus(upscanFileDescriptor: UpscanFileDescriptor): EitherT[Future, Error, Boolean]
+
+  def getAllUpscanCallBacks(cgtReference: CgtReference): EitherT[Future, Error, List[UpscanCallBack]]
+
+  def downloadFilesFromS3(
+    snapshot: UpscanSnapshot,
+    urls: List[UpscanCallBack]
+  ): EitherT[Future, Error, List[Either[Error, FileAttachment]]]
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala]: this return type is a bit weird - you've got errors in the right type as well. Would this not make more sense to be `EitherT[Future, Error, List[FileAttachment]]` and to collect the errors into one big one?

File: app/uk/gov/hmrc/cgtpropertydisposals/connectors/UpscanConnector.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  implicit executionContext: ExecutionContext
+) extends UpscanConnector
+    with Logging
+    with HttpErrorFunctions {
+
+  private lazy val userAgent: String = config.getConfString("appName", "cgt-property-disposal")
+
+  private val headers: Seq[(String, String)] = Seq(USER_AGENT -> userAgent)
+
+  implicit val contextShift: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/connectors/UpscanConnector.scala]: I'm going the `cats io` dep.

File: app/uk/gov/hmrc/cgtpropertydisposals/models/dms/DmsSubmissionPayload.scala [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.cgtpropertydisposals.models.dms
+
+final case class DmsSubmissionPayload(b64Html: B64Html, attachments: List[FileAttachment], dmsMetadata: DmsMetadata)
+//
+//object DmsSubmissionPayload {
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/models/dms/DmsSubmissionPayload.scala]: no, will be removed

File: app/uk/gov/hmrc/cgtpropertydisposals/models/name/IndividualName.scala [hunk 6]
```diff
@@ -28,3 +30,8 @@ object IndividualName {
 
   implicit val format: OFormat[IndividualName] = Json.format[IndividualName]
 }
+
+final case class Foo(s: LocalDateTime)
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/models/name/IndividualName.scala]: oops - will be removed.

File: app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala (hunk 2 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala]: `cats io` dep will be remvoed

File: app/uk/gov/hmrc/cgtpropertydisposals/connectors/UpscanConnector.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/connectors/UpscanConnector.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/models/dms/DmsSubmissionPayload.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/models/dms/DmsSubmissionPayload.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/models/name/IndividualName.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/models/name/IndividualName.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/service/DmsSubmissionService.scala (hunk 1 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/DmsSubmissionService.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala (hunk 2 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala (hunk 3 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala]: discussed this morning - not sure if it resolves to any other type than this - do you want to take a deeper look while I crack on with try to do integration testing?

Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala]: let's do that then, yeah. Let's get this thing tested

**Pull request merged**